### PR TITLE
Add verifiers environments

### PR DESCRIPTION
**Pull request by willccbb**
[Edit] Updated wandb logs:
- https://wandb.ai/primeintellect/reverse-text-debug/runs/6oshnx4o?nw=nwuserwillpi
- https://wandb.ai/primeintellect/reverse-text-debug/runs/90kvjygr?nw=nwuserwillpi

image

image

Fixed issue with prompt mask (was getting set to 1 instead of 0), curves basically match. We're still recomputing logprobs instead of using the vLLM ones -- the vLLM per-turn logprobs/tokens are available in `states` inside orchestrator, but we have to decide whether we want to reconstruct them while handling multi-turn tokenization edge cases vs. recompute.

simple-math: 

image
image

logs: https://wandb.ai/primeintellect/math-debug-wb?nw=nwuserwillpi

### CODE REVIEW

File: src/zeroband/training/environments/registry.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    dataset_name = env_args.get("dataset", "openai/gsm8k")
+    dataset = load_dataset(dataset_name, split="train")
+    vf_env = vf.SingleTurnEnv(dataset=dataset)
+    return vf_env
+
+
+REGISTRY = {
+    "gsm8k": load_gsm8k_environment,
+    "reverse-text": load_reverse_environment,
+    "default": load_default_environment,
```
Comment (mikasenghaas) [inline src/zeroband/training/environments/registry.py]: not sure i like `default` as a id here. directionally, i think we should have them be as descriptive as possible, i.e. just by reading the env ID, it's clear what tasks/ tools/ rewards it consists of. i understand this might be difficult. if that's the case, maybe registering an environment also entails a small description of the environment (docstring?).

the reason why i believe this is important is because in our current way of registering envs we will havea a wide diversity (some subset of the cross product of tasks, tools and rewards, and all possible compositions of that space), so I think we have to be a bit careful to keep structure here.

will, im sure you have some good ideas here

File: src/zeroband/training/orchestrator/config.py [hunk 2]
```diff
@@ -115,6 +100,12 @@ class DataConfig(BaseConfig):
     split: Annotated[str, Field(description="Split of the dataset to use.")] = "train"
 
 
+class EnvironmentConfig(BaseConfig):
+    """Configures the environment to be used for inference."""
+
+    id: Annotated[str, Field(description="ID of the environment to use.")] = "default"
```
Comment (mikasenghaas) [inline src/zeroband/training/orchestrator/config.py]: any reason we don't make this a list of ids and by default use the composed environment? could make naming easier, bc you don't have to find a new ID for a composed environment anymore

File: src/zeroband/training/orchestrator/config.py [hunk 3]
```diff
@@ -182,6 +173,9 @@ class OrchestratorConfig(BaseSettings):
     # The data configuration
     data: DataConfig = DataConfig()
```
Comment (mikasenghaas) [inline src/zeroband/training/orchestrator/config.py]: do we still need this?

File: src/zeroband/training/orchestrator/data.py [hunk 4]
```diff
@@ -32,18 +34,23 @@ def prepare_sample(
...
+    logprobs = torch.cat([torch.zeros(len(prompt_token_ids) - 1), torch.tensor(completion_logprobs)]).float()
     position_ids = torch.arange(total_tokens).long()
     advantages = torch.tensor(advantage).repeat(total_tokens).float()
 
     if total_tokens > seq_len:
         # We should never truncate as it would create a really bad learning signal. Instead, always set the maximum sequence length
         # on the inference worker accordingly, e.g. by setting the `max_tokens` parameter.
+        print(f"Prompt tokens: {prompt_tokens}")
+        print(f"Completion tokens: {completion_tokens}")
+        print(f"Sequence length: {seq_len}")
```
Comment (mikasenghaas) [inline src/zeroband/training/orchestrator/data.py]: Remove or use logger from `get_logger`

**Review by mikasenghaas — COMMENTED** [review]
this looks great already! can you include the w&b run of your reverse text repro

File: configs/training/reverse_text.toml [hunk 5]
```diff
@@ -2,6 +2,7 @@ max_steps = 30
 
 [model]
 name = "willcb/Qwen2.5-0.5B-Reverse-SFT"
+ac = true
```
Comment (samsja) [inline configs/training/reverse_text.toml]: I think that it should still run on 24gb without this

File: pyproject.toml [hunk 6]
```diff
@@ -28,6 +28,7 @@ dependencies = [
     "pyext",
     "reasoning_gym",
     "lovely-tensors>=0.1.18",
+    "verifiers",
```
Comment (samsja) [inline pyproject.toml:33]: lfg

File: src/zeroband/training/orchestrator/config.py (hunk 2 above)
Comment (samsja) [reply src/zeroband/training/orchestrator/config.py]: lets use non list for now for simplicity, goal is to merge asap

File: configs/training/reverse_text.toml (hunk 5 above)
Comment (willccbb) [reply configs/training/reverse_text.toml]: removed

File: src/zeroband/training/orchestrator/data.py [hunk 7]
```diff
@@ -32,18 +34,23 @@ def prepare_sample(
...
+    input_ids = torch.cat([prompt_token_ids, completion_token_ids]).long()
+    loss_mask = torch.cat([prompt_token_mask, completion_token_mask]).long()
     total_tokens = input_ids.shape[0]
-    logprobs = torch.cat([torch.zeros(len(input_tokens) - 1), torch.tensor(output_logprobs)]).float()
-    loss_mask = torch.cat([torch.zeros(len(input_tokens)), torch.ones(len(output_tokens))]).long()
+    logprobs = torch.cat([torch.zeros(len(prompt_token_ids) - 1), torch.tensor(completion_logprobs)]).float()
     position_ids = torch.arange(total_tokens).long()
     advantages = torch.tensor(advantage).repeat(total_tokens).float()
 
     if total_tokens > seq_len:
```
Comment (samsja) [inline src/zeroband/training/orchestrator/data.py]: we need to handle this case now as we can have sample that are going over seq_len. (need to impl overlng filtering here). Or is it done before ?

File: uv.lock [hunk 8]
```diff
@@ -1,13 +1,31 @@
...
     "platform_machine != 'aarch64' and sys_platform == 'linux'",
     "sys_platform == 'darwin'",
     "sys_platform != 'darwin' and sys_platform != 'linux'",
 ]
 
+[[package]]
+name = "accelerate"
+version = "1.8.1"
+source = { registry = "https://pypi.org/simple" }
+dependencies = [
```
Comment (samsja) [inline uv.lock]: do you know why accelerate got added ?

File: uv.lock [hunk 9]
```diff
@@ -1957,6 +1975,27 @@ wheels = [
     { url = "https://files.pythonhosted.org/packages/cb/40/1f922794af3dc7503f19319a8804b398a161a2cd54183cff8b12225b8d85/partial_json_parser-0.2.1.1.post6-py3-none-any.whl", hash = "sha256:abc332f09b13ef5233384dbfe7128a0e9ea3fa4b8f8be9b37ac1b433c810e99e", size = 10876, upload-time = "2025-06-23T17:51:44.332Z" },
 ]
 
+[[package]]
+name = "peft"
+version = "0.16.0"
+source = { registry = "https://pypi.org/simple" }
+dependencies = [
```
Comment (samsja) [inline uv.lock]: also we should not have peft

File: uv.lock [hunk 10]
```diff
@@ -3003,6 +3042,20 @@ wheels = [
     { url = "https://files.pythonhosted.org/packages/11/53/ce18470914ab6cfbec9384ee565d23c4d1c55f0548160b1c7b33000b11fd/triton-3.3.0-cp312-cp312-manylinux_2_27_x86_64.manylinux_2_28_x86_64.whl", hash = "sha256:b68c778f6c4218403a6bd01be7484f6dc9e20fe2083d22dd8aef33e3b87a10a3", size = 156504509, upload-time = "2025-04-09T20:27:40.413Z" },
 ]
 
+[[package]]
+name = "trl"
+version = "0.19.0"
+source = { registry = "https://pypi.org/simple" }
+dependencies = [
```
Comment (samsja) [inline uv.lock]: no trl

File: src/zeroband/training/environments/registry.py (hunk 1 above)
Comment (willccbb) [reply src/zeroband/training/environments/registry.py]: yeah i'd included this as a potential stopgap if we wanted to have simpler backwards compatibility with existing prime-rl tasks/rewards, but happy to deprecate

this will eventually be phased out by pihub, where environments will be loaded just like datasets, with arg options

e.g. 

```python
vf_env = pihub.load_env(env_id="primeintellect/reverse-text-debug", env_args={'num_samples': 1000})
```

File: src/zeroband/training/orchestrator/config.py (hunk 3 above)
Comment (willccbb) [reply src/zeroband/training/orchestrator/config.py]: removed

File: src/zeroband/training/orchestrator/data.py (hunk 7 above)
Comment (willccbb) [reply src/zeroband/training/orchestrator/data.py]: it's done before in process_env_results by setting `max_seq_length=config.seq_len`

```python
        results = vf_env.process_env_results(
            prompts=outputs["prompt"],
            completions=outputs["completion"],
            states=outputs["state"],
            rewards=outputs["reward"],
            processing_class=tokenizer,
            max_seq_length=config.seq_len or -1,
            mask_truncated_completions=True,  # TODO: make this configurable
            mask_env_responses=True,  # TODO: make this configurable
        )
```

Comment (willccbb) [reply src/zeroband/training/orchestrator/data.py]: made those both configurable in orchestrator config

File: uv.lock (hunk 8 above)
Comment (willccbb) [reply uv.lock]: will be fixed by uv.lock update; was previously mandatory in verifiers, now optional

File: uv.lock (hunk 9 above)
Comment (willccbb) [reply uv.lock]: yeah same thing with the verifiers dependencies haha

Comment (willccbb) [reply uv.lock]: fixed

File: uv.lock (hunk 8 above)
Comment (willccbb) [reply uv.lock]: fixed

File: src/zeroband/training/orchestrator/data.py (hunk 4 above)
Comment (willccbb) [reply src/zeroband/training/orchestrator/data.py]: removed

File: src/zeroband/training/orchestrator/config.py (hunk 2 above)
Comment (willccbb) [reply src/zeroband/training/orchestrator/config.py]: yeah there are a few other things to figure out re: composed environments + how we'll want to deal with this for pihub

File: uv.lock (hunk 10 above)
Comment (willccbb) [reply uv.lock]: fixed

File: src/zeroband/training/orchestrator/data.py (hunk 7 above)
Comment (samsja) [reply src/zeroband/training/orchestrator/data.py]: makes sense

File: src/zeroband/orchestrator/data.py [hunk 11]
```diff
@@ -32,12 +34,16 @@ def prepare_sample(
...
-    output_token_ids = torch.tensor(output_tokens).long()
-    input_ids = torch.cat([input_token_ids, output_token_ids]).long()
+    prompt_token_ids = torch.tensor(prompt_tokens).long()
+    prompt_token_mask = torch.tensor(prompt_mask).long()
+    completion_token_ids = torch.tensor(completion_tokens).long()
+    completion_token_mask = torch.tensor(completion_mask).long()
+    input_ids = torch.cat([prompt_token_ids, completion_token_ids]).long()
+
+    loss_mask = torch.cat([prompt_token_mask, completion_token_mask]).long()
+    # loss_mask = torch.cat([torch.zeros(len(prompt_tokens)), torch.ones(len(completion_tokens))]).long()
```
Comment (samsja) [inline src/zeroband/orchestrator/data.py]: can we remove the comments here

File: configs/trainer/simple_math_7b.toml
Comment (samsja) [inline configs/trainer/simple_math_7b.toml:1]: lets add hendrycks_math_7b then and intellect_math_7b

File: configs/orchestrator/intellect_math.toml [hunk 12]
```diff
@@ -1,22 +1,21 @@
...
 [monitor.wandb.log_samples]
 interval = 50
-num_samples = 16
\ No newline at end of file
+num_samples = 16
+
+# TODO: Bring back vf-based evals
+# [eval]
+# interval = 50
+# benchmarks = ["math500"]
```
Comment (samsja) [inline configs/orchestrator/intellect_math.toml]:
```suggestion
[eval]
interval = 50
benchmarks = ["math500"]
```

Comment (samsja) [reply configs/orchestrator/intellect_math.toml]: lets keep eval that way for now, we can merge vf base later

File: src/zeroband/orchestrator/orchestrator.py [hunk 13]
```diff
@@ -180,64 +158,96 @@ async def orchestrate(config: OrchestratorConfig):
...
+            rewards=outputs["reward"],
+            processing_class=tokenizer,
+            max_seq_length=config.seq_len or -1,
+            mask_truncated_completions=config.mask_truncated_completions,
+            mask_env_responses=config.mask_env_responses,
+        )
+        # TODO: use tokens/logprobs from vLLM request responses?
+        # states = outputs["state"]
+        # responses = [s["responses"][-1] for s in states] #s["responses"] is list of model turns
+        # completion_logprobs = [parse_logprobs(r) for r in responses]
```
Comment (samsja) [inline src/zeroband/orchestrator/orchestrator.py]: what is the conclusion here ?

Comment (willccbb) [reply src/zeroband/orchestrator/orchestrator.py]: to be determined later imo (if we want to reconstruct full token sequence + logprobs from vLLM outputs rather than recomputing, which will require resolving some multi-turn edge cases)

File: configs/orchestrator/intellect_math.toml (hunk 12 above)
Comment (mikasenghaas) [reply configs/orchestrator/intellect_math.toml]: ill bring it back

File: src/zeroband/orchestrator/orchestrator.py (hunk 13 above)
Comment (mikasenghaas) [reply src/zeroband/orchestrator/orchestrator.py]: we will try bringing it back, but in a later pr

File: src/zeroband/environments/registry.py [hunk 14]
```diff
@@ -0,0 +1,165 @@
...
+    from verifiers.utils.data_utils import extract_boxed_answer
+
+    from zeroband.orchestrator.genesys.math import compute_math_reward
+
+    train_dataset = load_dataset("PrimeIntellect/INTELLECT-2-only-math", split="train").map(
+        lambda x: {"question": x["prompt"], "info": json.loads(x["verification_info"]), "task": "simple-math"}
+    )
+    solve_rate_field = env_args.get("solve_rate_field", "solve_rate_qwen_r1_distill_7b")
+    min_solve_rate = env_args.get("min_solve_rate", 0.4)
+    max_solve_rate = env_args.get("max_solve_rate", 0.9)
```
Comment (samsja) [inline src/zeroband/environments/registry.py]: lets not do any filtering if the args are empty here
